user.py

```diff
@@ -113,11 +123,17 @@ def user_profile(user_id, current_year):
 
 @user_bp.route('/update_user_profile', methods=['POST'])
 @login_required
+@check_access_level(Role.ADMIN, Role.RESEARCHER, Role.TEACHER)
 def update_user_profile():
     form = request.form
     if not form:
         return make_response("Problem with form request", 500)
 
+    user_id = request.form['user_id']
```

Comment:
Beware, you fetch a string from the form and your `is_allowed_user` expects an int. I recommend you to set the user id in the route. This way, you don't need a hidden field in your form and just need to update the `action` attribute.